File: LeetCode_Algorithm/752(Open the Lock)/752.py
from typing import List
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        # sol1 970ms 58%
        # gerneral bfs

        # O(n^2 * A^n + D)
        # n은 다이얼의 개수 A는 해당하는 알파벳의 수
        # D는 deadend의 사이즈
        from collections import deque
        queue = deque([("0000", 0)])
        self.visited = set(deadends)
        while queue:
            num, cnt = queue.popleft()
            if num in self.visited:
                continue
            self.visited.add(num)
            if num == target:
                return cnt
            for i in range(4):
                one_plus = num[:i] + self.turnOneWheel(int(num[i]), 1) + num[i+1:]
                one_minus = num[:i] + self.turnOneWheel(int(num[i]), -1) + num[i+1:]
                queue.extend([(one_plus, cnt+1), (one_minus, cnt+1)])
        return -1
    def turnOneWheel(self, num, operand):
        num += operand
        if num == 10:
            return '0'
        if num == -1:
            return '9'
        return str(num)
        

        
    
print(Solution().openLock(["0201","0101","0102","1212","2002"], "0202"))
print(Solution().openLock(["8888"], "0009"))

File: LeetCode_Algorithm/752(Open the Lock)/752.js
var openLock = function(deadends, target) {
  // sol1 369ms 67%
  let visited = new Set(deadends)
  let queue = [["0000", 0]]
  while (queue.length) {
    [num, cnt] = queue.shift()
    if (num == target) {
      return cnt
    }
    if (visited.has(num)) {
      continue
    }
    visited.add(num)
    for (let i = 0 ; i < num.length; i++) {
      let clockwise = num.substring(0,i)+turnOneWheel(num[i], 1)+num.substring(i+1, )
      let antiClockwise = num.substring(0,i)+turnOneWheel(num[i], -1)+num.substring(i+1, )
      queue.push([clockwise, cnt+1],[antiClockwise, cnt+1])
    }
  }
  return -1
};

var turnOneWheel = function(n, operand) {
  n = Number(n)
  n += operand
  if (n == 10) {
    return '0'
  } else if (n == -1) {
    return '9'
  } else {
    return String(n)
  }
}
console.log(openLock(["0201","0101","0102","1212","2002"], "0202"))

